src/chart/layout/mod.rs

```diff
@@ -0,0 +1,995 @@
+            let end = self.axis_ranges.right.as_ref().unwrap().end;
+            for tick in axis.iter() {
+                // Find out where the tick is to be drawn.
+                let y_pos = scale_to_pixel(tick.pos, start, end, extent.y0, extent.y1);
+                draw_func((extent.x1, y_pos), tick, AxisSide::Right)?;
+            }
+        }
+        if let Some(axis) = suggestion.top {
+            let extent = self
+                .nodes
+                .get_chart_area_extent()
+                .ok_or_else(|| LayoutError::ExtentsError)?;
+            let start = self.axis_ranges.top.as_ref().unwrap().start;
+            let end = self.axis_ranges.top.as_ref().unwrap().end;
+            for tick in axis.iter() {
+                // Find out where the tick is to be drawn.
+                let x_pos = scale_to_pixel(tick.pos, start, end, extent.x0, extent.x1);
+                draw_func((x_pos, extent.y0), tick, AxisSide::Top)?;
+            }
+        }
+        if let Some(axis) = suggestion.bottom {
+            let extent = self
+                .nodes
+                .get_chart_area_extent()
+                .ok_or_else(|| LayoutError::ExtentsError)?;
+            let start = self.axis_ranges.bottom.as_ref().unwrap().start;
+            let end = self.axis_ranges.bottom.as_ref().unwrap().end;
+            for tick in axis.iter() {
+                // Find out where the tick is to be drawn.
+                let x_pos = scale_to_pixel(tick.pos, start, end, extent.x0, extent.x1);
+                draw_func((x_pos, extent.y1), tick, AxisSide::Bottom)?;
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Use some heuristics to guess the best tick spacing given the area we have.
+    fn suggest_tickmark_spacing_for_axes(
+        &self,
+    ) -> Result<AxisSpecs<SimpleLinearAxis<f32>>, DrawingAreaErrorKind<DB::ErrorType>> {
+        let da_extent = self.get_chart_area_extent()?;
+        let (w, h) = da_extent.size();
+        let ret = AxisSpecs {
+            top: self
+                .axis_ranges
+                .top
+                .as_ref()
+                .map(|range| suggest_tickmark_spacing_for_range(range, w)),
+            bottom: self
+                .axis_ranges
+                .bottom
+                .as_ref()
+                .map(|range| suggest_tickmark_spacing_for_range(range, w)),
+            left: self
+                .axis_ranges
+                .left
+                .as_ref()
+                .map(|range| suggest_tickmark_spacing_for_range(range, h)),
+            right: self
+                .axis_ranges
+                .right
+                .as_ref()
+                .map(|range| suggest_tickmark_spacing_for_range(range, h)),
+        };
+        Ok(ret)
+    }
+
+    /// Apply a cartesian grid to the chart area. Major/minor ticks are automatically
+    /// determined on a call to `draw`.
+    pub fn build_cartesian_2d(
```

Comment:
This function should be generic, because we need to support all types of coords. 

See https://github.com/38/plotters/blob/b82174aa1ac758ca6afb22a555d8732209fb2bb8/src/chart/builder.rs#L179

Reply:
I don't know how to turn `AsRangedCoord` into `Iterator<Tick>`, which is what my code needs to estimate the size of the tick labels.